### PR TITLE
#20479 Fix dead lock after SetDataAsync from clipboard in X11Clipboar…

### LINKED ISSUE #20479: [Linux] Fix dead lock after SetDataAsync from clipboard in X11Clipboard.cs
### Describe the bug

Call "Clipboard.SetTextAsync" will create a dead lock on UOS system.
UOS system is a Debian10-based Chinese system

### To Reproduce

1. Prepare a PC and a USB
2. Download system image from "https://www.chinauos.com/resource/download-professional"
3. Install Chinese UOS system (Debian 10) on your PC, you can choose an intel or amd cpu chipset
4. Try to use the call:

Dispatcher.UIThread.Post(async () =>
{
    var board = MainWindow.Inst.Clipboard;
    if (board != null)
        await board.SetTextAsync(text);    // Dead lock
});

### Expected behavior

board.SetTextAsync(text) could not cause a dead lock

### Avalonia version

11.3.11

### OS

Linux

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SweetLoverFT**
## What does the pull request do?

Solve a deadlock when call Linux X11 Clipboard to SetTextAsync

## What is the current behavior?

`board.SetTextAsync(text)` will create a dead lock on UOS system, e.g., UOS V20 1050/1060/1070

## What is the updated/expected behavior with this PR?

`board.SetTextAsync(text)` could not cause a dead lock

## How was the solution implemented (if it's not obvious)?

It's easy to reproduce but only on UOS System

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

Please check the code related with
```csharp
new TaskCompletionSource<T>()
```
must call `TrySetResult(T)` before return because there won't be a default value on UOS systems

## Obsoletions / Deprecations

Please check all code related with `new TaskCompletionSource<T>()`, must call `TrySetResult(T)` before return

## Fixed issues
Fixes #20479

### CODE REVIEW

**Comment by cla-avalonia** [comment]
- [x] All contributors have signed the CLA.

**Comment by SweetLoverFT** [comment]
@cla-avalonia agree

**Comment by SweetLoverFT** [comment]
I have confirmed that v12.0.999-cibuild0061523-alpha could not solve this issue.
```csharp
public void PasteText(string text)
{
    Dispatcher.UIThread.Post(async () =>
    {
        var board = MainWindow.Inst.Clipboard;
        if (board != null)
        {
            Logger.Inst.LogDebug($"text = {text}");
            await board.SetTextAsync(text);
            TypeWriter.Inst.CtrlCommand((byte)'v');
            Logger.Inst.LogDebug($"Done");
        }
    });
}
```
[uMouse.log](https://github.com/user-attachments/files/24708046/uMouse.log)

I wrote a speech recognition app and ignored the irrelevant logs, the app would use clipboard to copy the text transferred from websocket and paster it immediately, just like the code above. As you can see, it could generate the log before call
```csharp
await board.SetTextAsync(text);
```
However, no any logs `Done` will be generated after this call. Even worse, calling the clipboard caused a deadlock, resulting in the stuck and unresponsive GUI.

**Comment by SweetLoverFT** [comment]
**The root cause is about the C# `TaskCompletionSource<T>`**, let me explain in detail:
1. In fact, it will generate a C++ 11 thread and fire when you call it, like
```csharp
async Task StoreTextCoreAsync()
{
    // Skip storing atoms if the trivial formats are too big
    var text = await dataTransfer.TryGetTextAsync();
    if (text is null || text.Length * 2 > 64 * 1024)
        return;

    if (_storeAtomTcs is null || _storeAtomTcs.Task.IsCompleted)
        _storeAtomTcs = new TaskCompletionSource<bool>();

    var atoms = ConvertDataTransfer(dataTransfer);
    XChangeProperty(_x11.Display, _handle, _avaloniaSaveTargetsAtom, _x11.Atoms.XA_ATOM, 32,
        PropertyMode.Replace, atoms, atoms.Length);
    XConvertSelection(_x11.Display, _x11.Atoms.CLIPBOARD_MANAGER, _x11.Atoms.SAVE_TARGETS,
        _avaloniaSaveTargetsAtom, _handle, IntPtr.Zero);
    // _storeAtomTcs.TrySetResult(true);    // I add this
    await _storeAtomTcs.Task;
}
```
2. Once you call `_storeAtomTcs = new TaskCompletionSource<bool>();`, it will do like:
```cpp
auto t = std::thread(anonymousFunc/lambda);
t.detach();
```
3. The lambda will capture the below code with default return value in general, so it you don't set a return value, it's also correct on other Linux systems except UOS systems. Therefore, stuck and deadlock won't happen on other systems except for UOS.
4. before you `await xxx.Task`, if the `Task` is void or has been complete with returned value, it won't stuck at all. However, on UOS system, the generated code captured by the lambda won't generated any implicit return value at all, so `await _storeAtomTcs.Task;` will never get the returned value, this is the root cause, although it sounds like this is impossible because any C++ function, for example below, will have a default return value no matter whether you use `return` statement to assign it or not:
```cpp
bool funcA()
{
    // if you don't write a return statement, there will be a default value false
}
```
5. However, debian10-based UOS system will never generate a default return value, so the stuck happened.

**Comment by SweetLoverFT** [comment]
You can make an easy test by the code following:
```csharp
private async Task<bool> TestAsync()
{
    var tcs = new TaskCompletionSource<bool>();
    int a = 3, b = 4, c = a + b;
    await tcs.Task;    // stuck will happen
}

static int Main(string[] args)
{
    try
    {
        _ = TestAsync().ConfigureAwait(false).GetAwaiter().GetResult();
    }
    catch (Exception ex)
    {
        
    }
}
```
Therefore, I hope your team could check your code for each `TaskCompletionSource<T>` statement

Regards,
SweetLoverFT

**Comment by SweetLoverFT** [comment]
Currently, I have to use the following walkaround to solve this issue:
```csharp
public void PasteText(string text)
{
    Dispatcher.UIThread.Post(async () =>
    {
        var board = MainWindow.Inst.Clipboard;
        if (board != null)
        {
            var taskClipboard = board.SetTextAsync(text);
            try
            {
                // In fact, the copy paster was done, use a timeout task as walkaround
                await Task.WhenAny(taskClipboard, Task.Delay(TimeoutClipboard));
                TypeWriter.Inst.CtrlCommand((byte)'v');
            }
            catch (Exception ex)
            {

            }
        }
    });
}
```

**Comment by MrJul** [comment]
The whole point of a `TaskCompletionSource` is to be able to provide a `Task` that will be completed later.

In this case, the result is supposed to be set when a `SelectionClear` event is received: https://github.com/AvaloniaUI/Avalonia/blob/9c393b306ad30ccf36a49b33af814cfb5249356c/src/Avalonia.X11/Clipboard/X11Clipboard.cs#L57

~However, we don't currently check whether the selection owner of the `CLIPBOARD_MANAGER` is the Avalonia window. If it isn't, we won't ever receive any `SelectionClear`. That is _probably_ the correct fix, but I don't have UOS to test. Could you please try that?~
Edit: read the wrong method documentation, however we should get a `SelectionNotify` from the clipboard manager.

(Also note that technically, since is not a deadlock as only a single thread is involved.)

**Review by MrJul — CHANGES_REQUESTED** [review]
This is not a valid fix, see my comment above.

**Comment by SweetLoverFT** [comment]
> The whole point of a `TaskCompletionSource` is to be able to provide a `Task` that will be completed later.
> 
> In this case, the result is supposed to be set when a `SelectionClear` event is received:
> 
> https://github.com/AvaloniaUI/Avalonia/blob/9c393b306ad30ccf36a49b33af814cfb5249356c/src/Avalonia.X11/Clipboard/X11Clipboard.cs#L57
> 
> However, we don't currently check whether the selection owner of the `CLIPBOARD_MANAGER` is the Avalonia window. If it isn't, we won't ever receive any `SelectionClear`. That is _probably_ the correct fix, but I don't have UOS to test. Could you please try that?
> 
> (Also note that technically, since is not a deadlock as only a single thread is involved.)

Dear [MrJul](https://github.com/MrJul),

I got your point, there's no doubt that your description of Linux clipboard technology is entirely correct. Forgive me for phrasing, my expression may have misled you. What I meant was:
1. As for Linux clipboard, the code which is only involved with X11 is correct, so I said this only happened on UOS systems.
2. I have test the code only involved with X11 without the `TaskCompletionSource<T>()`, it won't block and could copy the data correctly
3. Error only occurred only with `TaskCompletionSource<T>()`

The root cause that I have pointed out, that is **UOS system cannot implicitly generate return values, this led to an endless wait**.
Therefore, you can regard this issue as an enhancement since it won't happen on other Linux systems, this is also why I wrote the code unrelated to the clipboard but involved with `TaskCompletionSource<T>()` as the unit for you.

This pull request is not particularly urgent, you can certainly choose not to fix it if you consider UOS to be a less important Linux distribution, I've already solved it with a different walkaround. However, **others similar code involved with `TaskCompletionSource<T>()` might encounter similar issues on UOS, this is what I want to share with you**.

My work around:
```csharp
public void PasteText(string text)
{
    Dispatcher.UIThread.Post(async () =>
    {
        var board = MainWindow.Inst.Clipboard;
        if (board != null)
        {
            var taskClipboard = board.SetTextAsync(text);
            try
            {
                // In fact, the copy & paste will always succeed with a timeout task as walkaround
                await Task.WhenAny(taskClipboard, Task.Delay(TimeoutClipboard));
                TypeWriter.Inst.CtrlCommand((byte)'v');
            }
            catch (Exception ex)
            {

            }
        }
    });
}
```

**Comment by MrJul** [comment]
> **UOS system cannot implicitly generate return values, this led to an endless wait**.

With all due respect, I think you have a deep misunderstanding of what is happening here. A `TaskCompletionSource` completes a `Task`. If such a mechanism wasn't working, you wouldn't be able to use anything async-related at all. There are no such things as "implicitly generated return values".

The only reason the task does not complete is because the code responsible for it (posted earlier) is never reached. It looks like your OS' clipboard manager does not take ownership of the clipboard for some reason.

I've tried an alternative fix with #20514, which should hopefully work better.

In any case, I'm going to close this PR as the proposed fix isn't valid, it's no different from returning an already completed task.

**Comment by SweetLoverFT** [comment]
I'm sure I understand `TaskCompletionSource`, these phenomenons happened only on **UOS/deepin** systems. All `TaskCompletionSource` can be used correctly when:
1. `TaskCompletionSource` doesn't need the return value, so it's correct
2. **As for `TaskCompletionSource<T>`, must call `TrySetResult(value)`** or other ways to assign a return value before return

The clipboard doesn't set the return value, which means
1. In fact, the clipboard has already done all of the copying actions except for `await Task`, you can write a simple console app to test
2. Then it will fail to get the return value, this led to an endless wait

I agree that you mark it as "close as no plan" or change the label "bug" to "enhancement" because actually UOS & deepin systems are a minority. And I'm sure you will find my statements are correct once you burn the system image and test avalonia on that.

a simple test for you once you have a UOS/deepin system machine
```csharp
using System;
using System.Threading.Tasks;

namespace TestTask
{
    internal class Program
    {
        private static async Task Test()
        {
            var tcs = new TaskCompletionSource<bool>();
            Console.WriteLine("Before declare a & b");
            int a = 3, b = 4;
            Console.WriteLine("Before c = a + b");
            int answer = a + b;
            Console.WriteLine($"answer = {a} + {b} = {answer}, Done");
            // tcs.TrySetResult(true);  // uncommon it to break the block
            await tcs.Task;
        }

        static void Main(string[] args)
        {
            try
            {
                // All WriteLine can be execute except below
                Test().ConfigureAwait(false).GetAwaiter().GetResult();
                Console.WriteLine("Can't arrived here unless called TrySetResult");
            }
            catch (Exception ex)
            {

            }
        }
    }
}
```

I hope Avalonia can become a great framework, so you will try this when you have time
